include/zephyr/sensing/sensing_sensor.h

```diff
@@ -0,0 +1,594 @@
+ * API support get separated sensitivity for each data field, or global
+ * sensitivity for all data fields.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The data buffer to receive value.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_get_sensitivity_t)(
+		const struct device *dev,
+		int index, uint32_t *value);
+
+/**
+ * @brief Set data range callback.
+ *
+ * Some sensors especially for physical sensors, support data range
+ * configuration, this may change data resolution.
+ *
+ * Since each sensor type may have multiple data fields for it's value, this
+ * API support set separated range for each data field, or global range for
+ * all data fields.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The value to be set.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_set_range_t)(
+		const struct device *dev,
+		int index, uint32_t value);
+
+/**
+ * @brief Get current data range callback.
+ *
+ * Some sensors especially for physical sensors, support data range
+ * configuration, this may change data resolution.
+ *
+ * Since each sensor type may have multiple data fields for it's value, this
+ * API support get separated range for each data field, or global range for
+ * all data fields.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param index The value fields index to be set, -1 for all fields (global).
+ *
+ * @param value The data buffer to receive value.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_get_range_t)(
+		const struct device *dev,
+		int index, uint32_t *value);
+
+/**
+ * @brief Set current sensor's hardware fifo size callback
+ *
+ * Some sensors especially for physical sensors, support hardware fifo, this API can
+ * configure the current fifo size.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param samples The sample number to set for fifo.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_set_fifo_t)(
+		const struct device *dev,
+		uint32_t samples);
+
+/**
+ * @brief Get current sensor's hardware fifo size
+ *
+ * Some sensors especially for physical sensors, support fifo, this API can
+ * get the current fifo size.
+ *
+ * @param dev The sensor instance device structure.
+ *
+ * @param samples The data buffer to receive the fifo sample number.
+ *
+ * @return 0 on success or negative error value on failure, not support etc.
+ */
+typedef int (*sensing_sensor_get_fifo_t)(
+		const struct device *dev,
+		uint32_t *samples);
```

Comment:
Can you explain why this is needed? These should be quite a level above needing hardware information (some might even be virtual sensors).

Reply:
two reasons get fifo callback is needed:
1) client sensor could be opened with less frequency than physical sensor, for example, physical accel is opened at 100HZ, while client sensor is opened at 1HZ, client sensor required to get 100 samples every 1s. If enable physical accel FIFO function, accel samples could be stored in FIFO, and every 1s, post all 100 samples to client. In this case, since client only be active every 1s, could reduce Sensing subsystem power.
2) Sensing subsystem working at polling mode, each sensor will process data with 10ms scheduling, If sensors support FIFO feature, it could work at higher frequency, such as 500HZ. physical sensor odr set as 500HZ, in each 10ms polling time, FIFO stored 5 samples, and every 10ms, post 5 samples to client.

Reply:
it's defined for later use, since FIFO mode should be a must in sensing subsystem, we do not plan to remove this function here. And we can implement FIFO function after you modified streaming sensor data in Kernel level driver later :)